sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/exemplar/ReservoirCell.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.exemplar;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.Clock;
+import io.opentelemetry.sdk.metrics.data.DoubleExemplarData;
+import io.opentelemetry.sdk.metrics.data.LongExemplarData;
+import io.opentelemetry.sdk.metrics.internal.data.ImmutableDoubleExemplarData;
+import io.opentelemetry.sdk.metrics.internal.data.ImmutableLongExemplarData;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * A Reservoir cell pre-allocated memories for Exemplar data.
+ *
+ * <p>We only allocate new objects during collection. This class should NOT cause allocations during
+ * sampling or within the synchronous metric hot-path.
+ *
+ * <p>Allocations are acceptable in the {@link #getAndResetDouble(Attributes)} and {@link
+ * #getAndResetLong(Attributes)} collection methods.
+ */
+class ReservoirCell {
+  private final Clock clock;
+  private long longValue;
```

Comment:
Ah hadn't thought of having both fields and both getAndReset methods, it seems to cut a lot of code without significant overhead (an extra few bytes per reservoir isn't that bad in practice). Let's add some comments that SDK instruments will only populate one of long or double depending on the type of the instrument.

Reply:
Considered a generic version of `ReservoirCell<T>` where `T` is either `Double` or `Long`, but opted for primitives to avoid the boxing / unboxing.